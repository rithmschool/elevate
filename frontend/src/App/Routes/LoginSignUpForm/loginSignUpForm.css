.LoginSignUpForm_login-or {
  position: relative;
  color: #aaa;
  margin-top: 10px;
  margin-bottom: 10px;
  padding-top: 10px;
  padding-bottom: 10px;
}
.LoginSignUpForm_span-or {
  display: block;
  position: absolute;
  left: 50%;
  top: -2px;
  margin-left: -25px;
  background-color: #fff;
  width: 50px;
  text-align: center;
}

.LoginSignUpForm_hr-or {
  height: 1px;
  margin-top: 0px !important;
  margin-bottom: 0px !important;
}

.LoginSignUpForm_link-signin,
.LoginSignUpForm_link-signup {
  background: none;
  border: none;
  padding: 0;
  font-family: arial, sans-serif;
  color: #069;
  text-decoration: underline;
  cursor: pointer;
}

.LoginSignUpForm_form-container {
  width: 30%;
  max-width: 450px;
}

.LoginSignUpForm_form-inside-container {
  padding: 2em;
}

@media screen and (max-width: 750px) {
  .LoginSignUpForm_form-container {
    width: 345px;
  }
}
