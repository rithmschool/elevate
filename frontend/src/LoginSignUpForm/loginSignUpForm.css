.login-or {
  position: relative;
  color: #aaa;
  margin-top: 10px;
  margin-bottom: 10px;
  padding-top: 10px;
  padding-bottom: 10px;
}
.span-or {
  display: block;
  position: absolute;
  left: 50%;
  top: -2px;
  margin-left: -25px;
  background-color: #fff;
  width: 50px;
  text-align: center;
}

.hr-or {
  height: 1px;
  margin-top: 0px !important;
  margin-bottom: 0px !important;
}
  
.button-signup, .button-signin {
  background: none!important;
  border: none;
  padding: 0!important;
  /*optional*/
  font-family: arial, sans-serif;
  /*input has OS specific font-family*/
  color: #069;
  text-decoration: underline;
  cursor: pointer;
}

.form-container {
  width: 30%;
  max-width: 450px;
}

.form-inside-container {
  padding: 2em;
}

.green-background {
  background-color:#98b398;
}

@media screen and (max-width: 750px) {
  .form-container {
    width: 345px;
  }
}

