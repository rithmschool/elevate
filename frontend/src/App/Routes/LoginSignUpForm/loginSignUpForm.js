import React from "react";
import { Button, Form } from "react-bootstrap";

import "./loginSignUpForm.css";
import ElevateApi from "../../../elevateApi";
import Spinner from "../../Spinner/spinner";
import LoginError from "./LoginError/loginError";

//created client_id from configure a project from google
const client_id =
  "98215850405-9u3oli17i7vko2f22k6rc7f9srlpjf3m.apps.googleusercontent.com";
let auth2;

class LoginSignUpForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      loginView: true,
      email: "",
      password: "",
      firstName: "",
      lastName: "",
      errors: [],
      isLoading: false
    };
  }

  /** To get profile information
   *  this code is from https://developers.google.com/identity/sign-in/web/sign-in
   */
  onSignIn = async googleUser => {
    // const profile = googleUser.getBasicProfile();
    // ID: profile.getId()
    // Name: profile.getName()
    // Email: profile.getEmail()

    /** this ID token will be sent to the server with HTTP post request
     *  in ElevateApi to get verify from google
     */
    var id_token = googleUser.getAuthResponse().id_token;

    /** wait to signinGoogle
     *  use google token to sign in to elevate api
     *  signinGoogle will return a token, use this token
     */
    let token;

    try {
      token = await ElevateApi.signinGoogle(id_token);
    } catch (errors) {
      return this.setState({ errors });
    }

    localStorage.setItem("token", token);
    await this.props.getCurrentUser();

    this.props.history.push("/dashboard");
  };

  componentDidMount() {
    /** Initialize the auth2 library and we use this auth2 with
     *  handleGoogleSignin below
     */
    try {
      window.gapi.load("auth2", () => {
        auth2 = window.gapi.auth2.init({
          client_id: client_id,
          fetch_basic_profile: false,
          scope: "profile email openid"
        });
      });
    } catch (errors) {
      return this.setState({ errors });
    }
  }

  /** auth2.signIn() gives back googleUser which can be used
   *  for argument in onSignIn method
   */
  handleGoogleSignin = async () => {
    this.onSignIn(await auth2.signIn());
  };

  loginOrSignup = evt => {
    if (evt.target.name === "login") {
      this.setState({ loginView: true });
    } else {
      this.setState({ loginView: false });
    }
  };

  handleChange = evt => {
    this.setState({ [evt.target.name]: evt.target.value });
  };

  handleSubmit = async evt => {
    evt.preventDefault();
    let token;

    this.setState({ isLoading: true });

    try {
      if (this.state.loginView) {
        const data = {
          email: this.state.email,
          password: this.state.password
        };

        token = await ElevateApi.login(data);
      } else {
        const data = {
          email: this.state.email,
          password: this.state.password,
          first_name: this.state.firstName,
          last_name: this.state.lastName
        };

        token = await ElevateApi.signup(data);
      }
    } catch (errors) {
      return this.setState({ isLoading: false, errors });
    }

    localStorage.setItem("token", token);

    await this.props.getCurrentUser();

    this.props.history.push("/dashboard");
  };

  render() {
    let loginView = this.state.loginView;

    if (this.state.isLoading) return <Spinner />;

    const firstLastNameInputs = (
      <div>
        <Form.Group>
          <Form.Control
            placeholder="First Name"
            className="signUpInput"
            id="firstName"
            name="firstName"
            type="text"
            onChange={this.handleChange}
            value={this.state.firstName}
          />
        </Form.Group>

        <Form.Group>
          <Form.Control
            placeholder="Last Name"
            className="signUpInput"
            id="lastName"
            name="lastName"
            type="text"
            onChange={this.handleChange}
            value={this.state.lastName}
          />
        </Form.Group>
      </div>
    );

    const emailPasswordInputs = (
      <div>
        <Form.Group>
          <Form.Control
            placeholder="Email"
            id="email"
            name="email"
            type="text"
            onChange={this.handleChange}
            value={this.state.email}
          />
        </Form.Group>

        <Form.Group>
          <Form.Control
            placeholder="Password"
            id="password"
            name="password"
            type="password"
            onChange={this.handleChange}
            value={this.state.password}
          />
        </Form.Group>
      </div>
    );

    const loginWithSocial = (
      <div>
        <div className="LoginSignUpForm_login-or">
          <hr className="LoginSignUpForm_hr-or" />
          <span className="LoginSignUpForm_span-or">or</span>
        </div>

        {/* Google Sign In Button */}
        <div className="row justify-content-center">
          <Button
            className="btn-block mr-3 ml-3"
            onClick={this.handleGoogleSignin}
          >
            <i className="fab fa-google"></i>
            <span>Sign in with Google</span>
          </Button>
        </div>
      </div>
    );

    const signUpButton = (
      <div className="row justify-content-center">
        <Button className="login-submit btn-block mr-3 ml-3" type="submit">
          Create an account
        </Button>
      </div>
    );

    const signInButton = (
      <div className="row justify-content-center">
        <Button className="login-submit btn-block mr-3 ml-3" type="submit">
          SIGN IN
        </Button>
      </div>
    );

    const directToSignIn = (
      <Form.Text id="signup" className="text-muted">
        <button
          name="login"
          className="LoginSignUpForm_link-signin"
          onClick={this.loginOrSignup}
        >
          Sign in
        </button>
      </Form.Text>
    );

    const directToSignUp = (
      <div>
        <Form.Text id="signup" className="text-muted mt-3">
          Don't have an account?
          <button
            className="LoginSignUpForm_link-signup"
            onClick={this.loginOrSignup}
          >
            Create One
          </button>
        </Form.Text>
      </div>
    );

    const forgotPasswordLink = (
      <div>
        <Form.Text id="signup" className="LoginSignUpForm_link-signup">
          <a href="http://localhost:3000/reset-password/forgot">
            Forgot password?
          </a>
        </Form.Text>
      </div>
    );

    return (
      <div
        className="
          container
          col-md-6
          offset-md-3
          col-lg-4
          offset-lg-4
          border 
          rounded 
          shadow"
      >
        <div className="LoginSignUpForm_form-inside-container mt-5">
          <Form onSubmit={this.handleSubmit}>
            <div className="mb-3">{loginView ? "Sign In" : "Sign Up"}</div>

            {emailPasswordInputs}
            {!loginView && firstLastNameInputs}
            {this.state.errors.length > 0 && <LoginError />}
            {loginView ? signInButton : signUpButton}
            {loginView ? loginWithSocial : loginWithSocial}
            {loginView && directToSignUp}
            {loginView ? forgotPasswordLink : directToSignIn}
          </Form>
        </div>
      </div>
    );
  }
}

export default LoginSignUpForm;
