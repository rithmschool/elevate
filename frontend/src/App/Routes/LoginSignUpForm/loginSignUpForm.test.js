import React from "react";
import { mount } from "enzyme";
import { render, fireEvent, act } from "@testing-library/react";
import toJson from "enzyme-to-json";
import LoginSignUpForm from "./loginSignUpForm";
import { Form } from "react-bootstrap";

describe("LoginSignUpForm", function() {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(<LoginSignUpForm />);
  });

  it("renders without crashing", function() {
    // Mounted in beforeEach above
  });

  it("matches snapshot", function() {
    const serialized = toJson(wrapper);

    expect(serialized).toMatchSnapshot();
  });

  it("renders 2 inputs on load", function() {
    expect(wrapper.find(Form.Control)).toHaveLength(2);
  });

  xit("fires loginorsignup function when 'create one' is clicked", async function() {});

  it("has the Google sign in button", function() {
    expect(wrapper.find("i.fa-google")).toHaveLength(1);
  });

  xit("Removes form and changes state to loading when submit is clicked", function() {
    // expect(wrapper.state().isLoading).toEqual(false);
    // const submit = wrapper.find("form").at(0);
    // submit.simulate("submit");
    // expect(wrapper.html()).toEqual('<div class="Spinner_loader"></div>');
    // expect(wrapper.state().isLoading).toEqual(true);
  });
});
