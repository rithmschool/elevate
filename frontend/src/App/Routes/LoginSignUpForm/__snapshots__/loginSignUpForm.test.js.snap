// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginSignUpForm matches snapshot 1`] = `
<LoginSignUpForm>
  <div
    className="
          container
          col-md-6
          offset-md-3
          col-lg-4
          offset-lg-4
          border 
          rounded 
          shadow"
    style={
      Object {
        "backgroundColor": "#F4F6F8",
        "marginTop": "10%",
      }
    }
  >
    <div
      className="LoginSignUpForm_form-inside-container mt-5"
    >
      <Form
        inline={false}
        onSubmit={[Function]}
      >
        <form
          className=""
          onSubmit={[Function]}
        >
          <div
            className="mb-3"
          >
            Sign In
          </div>
          <div>
            <FormGroup>
              <div
                className="form-group"
              >
                <FormControl
                  id="email"
                  name="email"
                  onChange={[Function]}
                  placeholder="Email"
                  type="text"
                  value=""
                >
                  <input
                    className="form-control"
                    id="email"
                    name="email"
                    onChange={[Function]}
                    placeholder="Email"
                    type="text"
                    value=""
                  />
                </FormControl>
              </div>
            </FormGroup>
            <FormGroup>
              <div
                className="form-group"
              >
                <FormControl
                  id="password"
                  name="password"
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
                >
                  <input
                    className="form-control"
                    id="password"
                    name="password"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
                  />
                </FormControl>
              </div>
            </FormGroup>
          </div>
          <div
            className="row justify-content-center"
          >
            <Button
              active={false}
              className="login-submit btn-block mr-3 ml-3"
              disabled={false}
              type="submit"
              variant="primary"
            >
              <button
                className="login-submit btn-block mr-3 ml-3 btn btn-primary"
                disabled={false}
                type="submit"
              >
                SIGN IN
              </button>
            </Button>
          </div>
          <div>
            <div
              className="LoginSignUpForm_login-or"
            >
              <hr
                className="LoginSignUpForm_hr-or"
              />
              <span
                className="LoginSignUpForm_span-or"
              >
                or
              </span>
            </div>
            <div
              className="row justify-content-center"
            >
              <Button
                active={false}
                className="btn-block mr-3 ml-3"
                disabled={false}
                onClick={[Function]}
                type="button"
                variant="primary"
              >
                <button
                  className="btn-block mr-3 ml-3 btn btn-primary"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <i
                    className="fab fa-google"
                  />
                  <span>
                    Sign in with Google
                  </span>
                </button>
              </Button>
            </div>
          </div>
          <div>
            <FormText
              className="text-muted mt-3"
              id="signup"
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              <small
                className="text-muted mt-3 form-text"
                id="signup"
                style={
                  Object {
                    "textAlign": "center",
                  }
                }
              >
                Don't have an account?
                <button
                  className="LoginSignUpForm_link-signup"
                  onClick={[Function]}
                >
                  Create One
                </button>
              </small>
            </FormText>
          </div>
          <div>
            <FormText
              className="LoginSignUpForm_link-signup"
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              <small
                className="LoginSignUpForm_link-signup form-text"
                style={
                  Object {
                    "textAlign": "center",
                  }
                }
              >
                <a
                  href="http://localhost:3000/reset-password/forgot"
                >
                  Forgot password?
                </a>
              </small>
            </FormText>
          </div>
        </form>
      </Form>
    </div>
  </div>
</LoginSignUpForm>
`;
